### ISSUE TITLE
allow decorators for functions

### ISSUE THREAD

**Issue by aleksey-bykov** [Microsoft/TypeScript#7318]
did a quick search, it looks like this hasn't been yet requested

currently decorators can only be added to classes, methods and properties

consider allowing them on functions too

**Comment by tinganho** [comment]
@aleksey-bykov it was discussed in here https://github.com/Microsoft/TypeScript/issues/2249#issuecomment-77690164

**Comment by mhegazy** [comment]
Please note that decorator proposal is undergoing churn now, and we would not be adding new features until we have a stable proposal.

**Comment by IgorMinar** [comment]
@mhegazy just fyi that function expression decorators are at stage 0: https://github.com/tc39/proposals/blob/master/stage-0-proposals.md

And while the underlying implementation is undergoing some churn, the syntax has been stable for quite a while. It would be great to have this implemented behind a flag because angular users could then write tests with a much nicer syntax.

```
it('should do foo', @Inject()(fooService:FooService) => {
...
});
```

compared to the current:

```
it('should do foo', inject([FooService], (fooService:FooService) => {
...
});
```

**Comment by majo44** [comment]
Any progress ? This will be really powerful feature. There is a lot of use cases.   
I see on roadmap just support for function expression/arrow functions :( 
I know the problem is with function hoisting. 
But I think we can loos hoisting for decorated functions (especially if we are using modules).
Compiler is able to detect decorated function hoisting (and cycles) and just throw error.
I beg, do not give up  :)

Related discussion:
[wycats/javascript-decorators/Allow decorators for functions as well](https://github.com/wycats/javascript-decorators/issues/4)

**Comment by Thinkscape** [comment]
This could potentially solve #10644 - it's a nasty side-effect when trying to use using bound methods feature with decorators :-(

**Comment by aluanhaddad** [comment]
My understanding of this issue is:
Decorators were added due mitigate the lack of expressiveness that was inherent in the _maximally minimal_ approach to classes that TC39 took.

My crude paraphrasing of the above is:
_ES2015_ classes are basically useless without decorators. They are a travesty of a class implementation.

Do they offer conciseness? No.
Do they enable expressing anything that is not better expressed with factories? Not for me.

**Comment by aluanhaddad** [comment]
Wow, that was a fast 3 👎.

I am not saying decorators should not be allowed on functions. I am saying that the reason they are not is because they were only added to make up for classes being useless.

I would like to have function decorators.

**Comment by zixia** [comment]
I beg, do not give up +1

**Comment by hax** [comment]
The standardization effort of function decorators has been moved to https://github.com/iddan/proposal-function-expression-decorators . The hoisting issue is discussed in No. 3 issue.

**Comment by OmgImAlexis** [comment]
Is there any news on this? Looks like nothing really changed in the last year or so.

**Comment by marco-jantke** [comment]
AFAIU this is still blocked by the lack of JavaScripts ability to do that.

**Comment by Exeteres** [comment]
Hi there!
I just created a simple [plugin](https://github.com/Exeteres/function-decorator-ts-plugin) that enables function decorators.
It is a system of two plugins: `tsserver` plugin that removes errors and `ts-patch` (`ttypescript`) plugin that transforms decorated function to the corresponding expression.
<details>
<summary>Example</summary>

![image](https://user-images.githubusercontent.com/29506045/86827555-54f5c900-c0bc-11ea-9d36-dd55cd0c05ea.png)
![image](https://user-images.githubusercontent.com/29506045/86827585-60e18b00-c0bc-11ea-8b0d-3ddd1eb63da9.png)

</details>

For [some reason](https://github.com/cevek/ttypescript/issues/99), this plugin can only be used with `ts-patch`, not with `ttypescript`. 
And of course there are some [limitations](https://github.com/Exeteres/function-decorator-ts-plugin#limitations).

**Comment by andreujuanc** [comment]
If the argument is that this will break hoisting, GOOD THEN. I hate hoisting anyways 😆 

Jokes aside, the fact that decorators can only be used on classes is totally arbitrary. Everything is possible. Even a dangling decorator are technically possible to implement. This thread is more like convincing the right people to actually make it happen.

**Comment by bal1anD** [comment]
Any updates on this? This will be a very valuable feature

**Comment by MartinJohns** [comment]
@bal1anD Just check yourself:

> The standardization effort of function decorators has been moved to https://github.com/iddan/proposal-function-expression-decorators .

The proposal is still stage 0. It builds upon the decorator porposal, which is still only stage 2. Don't expect this anytime soon.

**Comment by menocomp** [comment]
Using decorators in functions!  Why?
We can use higher order functions that is even much better and handle more use cases that decorators can not handle.

**Comment by bal1anD** [comment]
@menocomp: well, in that case, why introduce decorators in TS at all? All of their use-cases can be solved by higher-order functions? The reason is elegance, clean-code.

**Comment by andreujuanc** [comment]
@bal1anD  not only that but some people find this feature useful 
![image](https://user-images.githubusercontent.com/360052/159001063-5db9c25e-0530-4e10-8cde-73090cf45c02.png)

**Comment by vegerot** [comment]
This would be helpful as well

**Comment by aspirisen** [comment]
Decorators for function would be extremely useful for react, so instead of this:
```
export const Component = forwardRef(function Component() {
    return <div>Component</div>
})

// or

function Component() {
    return <div>Box</div>
}

const _Component = forwardRef(Component)

export { _Component as Component }

```

We could have just
```
@forwardRef
function Component() {
    return <div>Component</div>
}
```

**Comment by nowaysgit** [comment]
+ 1 !

**Comment by tareksaleem** [comment]
I think this could be done using a hack
instead of waiting for js to support decorators or change things in hoisting we do something different
basically all what is needed from typescript is supporting the syntax itself and this syntax could be transpilled to javascript syntax normally, how this could be done?

aft first anonymous functions could not be supported so let's stick with functions and suppose we have this function:

```typescript
@logger
function fetchData() {
  await fetch('resource')
}
```

logger actually will not be a class or method decorator to avoid dealing with descriptors, instead of that it will be a higher order function as the following:
```typescript
function logger(target) {
  return (...args) {
    // do whatever you want
    return target(...args)
  }
}
```
when typescript compile the code it should output something like this:

```typescript
function fetchData() {
   logger()
  // ...other decorators
  return _fetchData();
}
```

I think this could be a start

**Comment by MartinJohns** [comment]
> I think this could be done using a hack
instead of waiting for js to support decorators

Auch hacks are explicitly out of scope for TypeScript.

**Comment by brillout** [comment]
There is a new(-ish) TC39 repo for function decorators:
 - https://github.com/tc39/proposal-function-and-object-literal-element-decorators
 - https://github.com/tc39/proposal-function-and-object-literal-element-decorators/issues/4
 - https://github.com/tc39/proposal-function-and-object-literal-element-decorators/issues/5

Since TypeScript adoption is sufficient for being able to use new language syntax, I wonder if experimenting function decorators directly with the TypeScript team would be an option to fasten the process.

**Comment by menocomp** [comment]
10 years passed so far.
I honestly not sure why this issue is raised here in typescript instead of introducing it in javascript language first.
